pkg/store/cache.go

```diff
@@ -1,47 +1,159 @@
 package store
 
 import (
-	lru "github.com/hashicorp/golang-lru"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru/simplelru"
 	"github.com/oklog/ulid"
+	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/tsdb/labels"
 )
 
+const (
+	cacheTypePostings = "postings"
+	cacheTypeSeries   = "series"
+)
+
 type cacheItem struct {
 	block ulid.ULID
 	key   interface{}
 }
+
+func (c cacheItem) keyType() string {
+	switch c.key.(type) {
+	case cacheKeyPostings:
+		return cacheTypePostings
+	case cacheKeySeries:
+		return cacheTypeSeries
+	}
+	return "<unknown>"
+}
+
 type cacheKeyPostings labels.Label
 type cacheKeySeries uint64
 
 type indexCache struct {
-	lru *lru.TwoQueueCache
+	mtx     sync.Mutex
+	lru     *lru.LRU
+	maxSize uint64
+	curSize uint64
+
+	requests    *prometheus.CounterVec
+	hits        *prometheus.CounterVec
+	added       *prometheus.CounterVec
+	current     *prometheus.GaugeVec
+	currentSize *prometheus.GaugeVec
 }
 
-func newIndexCache(sz int) (*indexCache, error) {
-	l, err := lru.New2Q(sz)
+// newIndexCache creates a new LRU cache for index entries and ensures the total cache
+// size approximately does not exceed maxBytes.
+func newIndexCache(reg prometheus.Registerer, maxBytes uint64) (*indexCache, error) {
+	c := &indexCache{
+		maxSize: maxBytes,
+	}
+	evicted := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_store_index_cache_items_evicted_total",
+		Help: "Total number of items that were evicted from the index cache.",
+	}, []string{"item_type"})
+
+	c.added = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_store_index_cache_items_added_total",
+		Help: "Total number of items that were added to the index cache.",
+	}, []string{"item_type"})
+
+	c.requests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_store_index_cache_requests_total",
+		Help: "Total number of requests to the cache.",
+	}, []string{"item_type"})
+
+	c.hits = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_store_index_cache_hits_total",
+		Help: "Total number of requests to the cache that were a hit.",
+	}, []string{"item_type"})
+
+	c.current = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "thanos_store_index_cache_items",
+		Help: "Current number of items in the index cache.",
+	}, []string{"item_type"})
+
+	c.currentSize = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "thanos_store_index_cache_items_size_bytes",
+		Help: "Current byte size of items in the index cache.",
+	}, []string{"item_type"})
+
+	// Initialize LRU cache with a high size limit since we will manage evictions ourselves
+	// based on stored size.
+	l, err := lru.NewLRU(1e12, func(key, val interface{}) {
+		k := key.(cacheItem).keyType()
+		v := val.([]byte)
+
+		evicted.WithLabelValues(k).Inc()
+		c.current.WithLabelValues(k).Dec()
+		c.currentSize.WithLabelValues(k).Sub(float64(len(v)))
+
+		c.curSize -= uint64(len(v))
+	})
```

Comment:
This is the eviction callback of the LRU implementation.

Reply:
Worth to add a comment or construct the function in separate line with the name then (: